Reuse parsing protocolSupportEnumeration for AttributeAuthorityDescriptor

vladimir-mencl-eresearch opened the pull request:
Hi @btmattsson ,

Many thanks for refactoring the validation code.

I have now rolled that out in my testbed and this PR fixes some wrinkles I bumped into.

Main one was that for an IdP with an `AttributeAuthorityDescriptor`, I was getting Error:
> saml-bindings-2.0-os: SAML2 Binding in md:AttributeService[Binding=urn:oasis:names:tc:SAML:2.0:bindings:SOAP], but SAML2 not supported in AttributeAuthorityDescriptor.
even though the metadata was correct.

I found that the `parseAttributeAuthorityDescriptor` function was disconnected from how `protocolSupportEnumeration` was parsed in `parseIDPSSODescriptorparseIDPSSODescriptor`.

I have refactored that code into a reusable function `parseProtocolSupportEnumeration` that's now used by all the three `parse*Descriptor` functions.

As the results from the parsing are specific to the role descriptor parsed, I removed the top-level attributes of the `ParseXML` class holding the results, and these are stored only in local variables in the `parse*Descriptor` methods.

Unfortunately, in the `ParseXMLSWAMID` sub-class, I had to repeat the call to `parseProtocolSupportEnumeration` - could not see a good way to pass the result between the generic and federation-specific parse methods.

Please let me know if happy with this approach.

Cheers,
Vlad

vladimir-mencl-eresearch commented:
@btmattsson , if you are OK with the message change in 884a728a12, I'd do a similar change for the remaining messages in `git grep 'missing in validator'`

I see this pattern repeating: a `switch` on `($child->nodeName)`, with the `default` branch of the `switch` reporting the node as "missing" ... while I understand the code is really trying to say it found an unknown node type it cannot handle.

Hence I think we should change the message to report is as `Unknown <something> found`.

Please let me know if you're OK with me making this change across the other locations in `src/ParseXML.php`.

Cheers,
Vlad

btmattsson commented:
> Hi @btmattsson ,
> 
> Many thanks for refactoring the validation code.
> 
> I have now rolled that out in my testbed and this PR fixes some wrinkles I bumped into.
> 
> Main one was that for an IdP with an `AttributeAuthorityDescriptor`, I was getting Error:
> 
> > saml-bindings-2.0-os: SAML2 Binding in md:AttributeService[Binding=urn:oasis:names:tc:SAML:2.0:bindings:SOAP], but SAML2 not supported in AttributeAuthorityDescriptor.
> > even though the metadata was correct.
> 
> I found that the `parseAttributeAuthorityDescriptor` function was disconnected from how `protocolSupportEnumeration` was parsed in `parseIDPSSODescriptorparseIDPSSODescriptor`.
> 
> I have refactored that code into a reusable function `parseProtocolSupportEnumeration` that's now used by all the three `parse*Descriptor` functions.
> 
> As the results from the parsing are specific to the role descriptor parsed, I removed the top-level attributes of the `ParseXML` class holding the results, and these are stored only in local variables in the `parse*Descriptor` methods.
> 
> Unfortunately, in the `ParseXMLSWAMID` sub-class, I had to repeat the call to `parseProtocolSupportEnumeration` - could not see a good way to pass the result between the generic and federation-specific parse methods.
> 
> Please let me know if happy with this approach.
> 
> Cheers, Vlad

The reason for having them global and making them false in every parse was that then also extended class function could reach them.

vladimir-mencl-eresearch commented:
> The reason for having them global and making them false in every parse was that then also extended class function could reach them.

I saw that ... but they're only valid within the context of of a role descriptor.

If made global, there would have to be separate values kept for each roleDescriptor.

Either
```
$idpssoSaml2Found = false;
...
```
or
```
$samlRolesFound = array(
  `IDPSSODescriptor` => array(
       `saml2` => fa;se,
....
```

Would either of these look reasonable to you?

Cheers,
Vlad

btmattsson commented:
> > The reason for having them global and making them false in every parse was that then also extended class function could reach them.
> 
> I saw that ... but they're only valid within the context of of a role descriptor.
> 
> If made global, there would have to be separate values kept for each roleDescriptor.
> 
> Either
> 
> ```
> $idpssoSaml2Found = false;
> ...
> ```
> 
> or
> 
> ```
> $samlRolesFound = array(
>   `IDPSSODescriptor` => array(
>        `saml2` => fa;se,
> ....
> ```
> 
> Would either of these look reasonable to you?
> 
> Cheers, Vlad

For me the option with

 ```
$samlRolesFound = array(
   `IDPSSODescriptor` => array(
        `saml2` => false,
 ....
 ```

looks most attractive.

// Björn M.

vladimir-mencl-eresearch commented:
Hi @btmattsson ,

I've just reworked the PR to use global state for parsed protocolSupportEnumeration results.  It will definitely make it easier to use these results in methods of subclasses.

I have also updated the `missing in validator` messages to `Unknown element found`.  Please let me know if OK with this change.  Happy to revert it if you'd rather keep it that way.  (I only now grasped the meaning of "<support for this element> missing in validator" - conservative approach not to call out things in metadata that are not implemented yet).  Please let me know which way you'd like to go on this.

Cheers,
Vlad
